File: frontend/lib/regionDetect.ts
export const ALL_AWS_REGIONS = [
  "us-east-1",
  "us-east-2",
  "us-west-1",
  "us-west-2",
  "ca-central-1",
  "eu-west-1",
  "eu-west-2",
  "eu-west-3",
  "eu-central-1",
  "eu-north-1",
  "ap-southeast-1",
  "ap-southeast-2",
  "ap-northeast-1",
  "ap-northeast-2",
  "ap-northeast-3",
  "ap-south-1",
  "sa-east-1",
  "me-south-1",
  "af-south-1",
] as const;

export type AWSRegion = (typeof ALL_AWS_REGIONS)[number];

export const REGION_LABELS: Record<string, string> = {
  "us-east-1": "N. Virginia",
  "us-east-2": "Ohio",
  "us-west-1": "N. California",
  "us-west-2": "Oregon",
  "ca-central-1": "Canada",
  "eu-west-1": "Ireland",
  "eu-west-2": "London",
  "eu-west-3": "Paris",
  "eu-central-1": "Frankfurt",
  "eu-north-1": "Stockholm",
  "ap-southeast-1": "Singapore",
  "ap-southeast-2": "Sydney",
  "ap-northeast-1": "Tokyo",
  "ap-northeast-2": "Seoul",
  "ap-northeast-3": "Osaka",
  "ap-south-1": "Mumbai",
  "sa-east-1": "Sao Paulo",
  "me-south-1": "Bahrain",
  "af-south-1": "Cape Town",
};

const REGION_COORDINATES: Record<AWSRegion, { lat: number; lon: number }> = {
  "us-east-1": { lat: 39.0438, lon: -77.4874 },
  "us-east-2": { lat: 40.4173, lon: -82.9071 },
  "us-west-1": { lat: 37.3382, lon: -121.8863 },
  "us-west-2": { lat: 45.5231, lon: -122.6765 },
  "ca-central-1": { lat: 45.5017, lon: -73.5673 },
  "eu-west-1": { lat: 53.3498, lon: -6.2603 },
  "eu-west-2": { lat: 51.5074, lon: -0.1278 },
  "eu-west-3": { lat: 48.8566, lon: 2.3522 },
  "eu-central-1": { lat: 50.1109, lon: 8.6821 },
  "eu-north-1": { lat: 59.3293, lon: 18.0686 },
  "ap-southeast-1": { lat: 1.3521, lon: 103.8198 },
  "ap-southeast-2": { lat: -33.8688, lon: 151.2093 },
  "ap-northeast-1": { lat: 35.6762, lon: 139.6503 },
  "ap-northeast-2": { lat: 37.5665, lon: 126.978 },
  "ap-northeast-3": { lat: 34.6937, lon: 135.5023 },
  "ap-south-1": { lat: 19.076, lon: 72.8777 },
  "sa-east-1": { lat: -23.5505, lon: -46.6333 },
  "me-south-1": { lat: 26.0667, lon: 50.5577 },
  "af-south-1": { lat: -33.9249, lon: 18.4241 },
};

const TIMEZONE_TO_REGIONS: Record<string, AWSRegion[]> = {
  America: ["us-east-1", "us-west-2", "ca-central-1"],
  US: ["us-east-1", "us-west-2", "us-east-2"],
  Canada: ["ca-central-1", "us-east-1", "us-west-2"],
  Europe: ["eu-central-1", "eu-west-1", "eu-north-1"],
  Africa: ["af-south-1", "eu-west-1", "me-south-1"],
  Asia: ["ap-southeast-1", "ap-northeast-1", "ap-south-1"],
  Australia: ["ap-southeast-2", "ap-southeast-1", "ap-northeast-1"],
  Pacific: ["ap-southeast-2", "ap-northeast-1", "us-west-2"],
  Atlantic: ["eu-west-1", "us-east-1", "sa-east-1"],
  Indian: ["ap-south-1", "ap-southeast-1", "me-south-1"],
};

const CITY_OVERRIDES: Record<string, AWSRegion[]> = {
  "America/New_York": ["us-east-1", "us-east-2", "ca-central-1"],
  "America/Chicago": ["us-east-1", "us-west-2", "us-east-2"],
  "America/Denver": ["us-west-2", "us-east-1", "us-west-1"],
  "America/Los_Angeles": ["us-west-2", "us-west-1", "us-east-1"],
  "America/Sao_Paulo": ["sa-east-1", "us-east-1", "eu-west-1"],
  "Europe/London": ["eu-west-2", "eu-west-1", "eu-central-1"],
  "Europe/Berlin": ["eu-central-1", "eu-west-1", "eu-north-1"],
  "Europe/Paris": ["eu-west-3", "eu-central-1", "eu-west-1"],
  "Europe/Stockholm": ["eu-north-1", "eu-central-1", "eu-west-1"],
  "Asia/Tokyo": ["ap-northeast-1", "ap-northeast-3", "ap-northeast-2"],
  "Asia/Seoul": ["ap-northeast-2", "ap-northeast-1", "ap-northeast-3"],
  "Asia/Singapore": ["ap-southeast-1", "ap-northeast-1", "ap-south-1"],
  "Asia/Kolkata": ["ap-south-1", "ap-southeast-1", "me-south-1"],
  "Asia/Dubai": ["me-south-1", "ap-south-1", "eu-central-1"],
  "Australia/Sydney": ["ap-southeast-2", "ap-southeast-1", "ap-northeast-1"],
};

const DEFAULT_REGIONS: AWSRegion[] = ["us-east-1", "us-west-2", "eu-west-1"];

type IpLookupOptions = {
  fetcher?: typeof fetch;
  fallbackTimezone?: string;
};

function toRadians(value: number): number {
  return (value * Math.PI) / 180;
}

function distanceInKm(fromLat: number, fromLon: number, toLat: number, toLon: number): number {
  const earthRadiusKm = 6371;
  const dLat = toRadians(toLat - fromLat);
  const dLon = toRadians(toLon - fromLon);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRadians(fromLat)) * Math.cos(toRadians(toLat)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return earthRadiusKm * c;
}

function asNumber(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string") {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return null;
}

function fallbackRegions(timezone?: string): AWSRegion[] {
  return detectRecommendedRegions(timezone ?? detectTimezone());
}

export function detectClosestRegionsFromCoordinates(lat: number, lon: number, limit = 3): AWSRegion[] {
  return [...ALL_AWS_REGIONS]
    .sort((a, b) => {
      const aCoordinates = REGION_COORDINATES[a];
      const bCoordinates = REGION_COORDINATES[b];
      const distanceA = distanceInKm(lat, lon, aCoordinates.lat, aCoordinates.lon);
      const distanceB = distanceInKm(lat, lon, bCoordinates.lat, bCoordinates.lon);
      return distanceA - distanceB;
    })
    .slice(0, Math.max(1, limit));
}

export function detectRecommendedRegions(timezone?: string): AWSRegion[] {
  if (!timezone) return DEFAULT_REGIONS;

  const cityMatch = CITY_OVERRIDES[timezone];
  if (cityMatch) return cityMatch;

  const continent = timezone.split("/")[0];
  const continentMatch = TIMEZONE_TO_REGIONS[continent];
  if (continentMatch) return continentMatch;

  return DEFAULT_REGIONS;
}

export function detectTimezone(): string | undefined {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone;
  } catch {
    return undefined;
  }
}

export async function detectRecommendedRegionsByIp(options: IpLookupOptions = {}): Promise<AWSRegion[]> {
  const fallback = fallbackRegions(options.fallbackTimezone);
  const fetcher = options.fetcher ?? fetch;

  try {
    const response = await fetcher("https://ipapi.co/json/");
    if (!response.ok) return fallback;

    const payload = (await response.json().catch(() => null)) as Record<string, unknown> | null;
    if (!payload) return fallback;

    const latitude = asNumber(payload.latitude ?? payload.lat);
    const longitude = asNumber(payload.longitude ?? payload.lon);
    if (latitude === null || longitude === null) return fallback;

    return detectClosestRegionsFromCoordinates(latitude, longitude);
  } catch {
    return fallback;
  }
}

File: backend/region_detect.py
import asyncio
import json
import math
from urllib.error import URLError
from urllib.request import urlopen

_DEFAULT_REGION = "us-east-1"
_IPAPI_TIMEOUT_SECONDS = 3.0

_REGION_COORDINATES: dict[str, tuple[float, float]] = {
    "us-east-1": (39.0438, -77.4874),
    "us-east-2": (40.4173, -82.9071),
    "us-west-1": (37.3382, -121.8863),
    "us-west-2": (45.5231, -122.6765),
    "ca-central-1": (45.5017, -73.5673),
    "eu-west-1": (53.3498, -6.2603),
    "eu-west-2": (51.5074, -0.1278),
    "eu-west-3": (48.8566, 2.3522),
    "eu-central-1": (50.1109, 8.6821),
    "eu-north-1": (59.3293, 18.0686),
    "ap-southeast-1": (1.3521, 103.8198),
    "ap-southeast-2": (-33.8688, 151.2093),
    "ap-northeast-1": (35.6762, 139.6503),
    "ap-northeast-2": (37.5665, 126.9780),
    "ap-northeast-3": (34.6937, 135.5023),
    "ap-south-1": (19.0760, 72.8777),
    "sa-east-1": (-23.5505, -46.6333),
    "me-south-1": (26.0667, 50.5577),
    "af-south-1": (-33.9249, 18.4241),
}


def _to_radians(value: float) -> float:
    return (value * math.pi) / 180


def _distance_in_km(from_lat: float, from_lon: float, to_lat: float, to_lon: float) -> float:
    earth_radius_km = 6371
    d_lat = _to_radians(to_lat - from_lat)
    d_lon = _to_radians(to_lon - from_lon)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(_to_radians(from_lat))
        * math.cos(_to_radians(to_lat))
        * math.sin(d_lon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return earth_radius_km * c


def _as_float(value: object) -> float | None:
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return None
    return None


def _fetch_location_coordinates(client_ip: str | None) -> tuple[float, float] | None:
    endpoint = (
        f"https://ipapi.co/{client_ip}/json/"
        if isinstance(client_ip, str) and client_ip.strip()
        else "https://ipapi.co/json/"
    )

    try:
        with urlopen(endpoint, timeout=_IPAPI_TIMEOUT_SECONDS) as response:
            payload = json.loads(response.read().decode("utf-8"))
    except (URLError, TimeoutError, json.JSONDecodeError, ValueError):
        return None

    if not isinstance(payload, dict):
        return None

    latitude = _as_float(payload.get("latitude") or payload.get("lat"))
    longitude = _as_float(payload.get("longitude") or payload.get("lon"))
    if latitude is None or longitude is None:
        return None

    return latitude, longitude


def _closest_region(latitude: float, longitude: float) -> str:
    closest_code = _DEFAULT_REGION
    closest_distance = float("inf")
    for code, (region_lat, region_lon) in _REGION_COORDINATES.items():
        distance = _distance_in_km(latitude, longitude, region_lat, region_lon)
        if distance < closest_distance:
            closest_distance = distance
            closest_code = code
    return closest_code


async def detect_closest_region(client_ip: str | None) -> str:
    coordinates = await asyncio.to_thread(_fetch_location_coordinates, client_ip)
    if coordinates is None:
        return _DEFAULT_REGION
    return _closest_region(coordinates[0], coordinates[1])
